odohlistener.go

```diff
@@ -0,0 +1,250 @@
+		slog.String("client", r.RemoteAddr),
+		slog.String("target", host),
+	)
+	response, err := forwardProxyRequest(s.proxyClient, host, path, b, contentType)
+	if err != nil {
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+	if response.StatusCode != 200 {
+		http.Error(w, http.StatusText(response.StatusCode), response.StatusCode)
+		return
+	}
+
+	defer response.Body.Close()
+	rb, err := io.ReadAll(response.Body)
+	if err != nil {
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+	w.Header().Set("Content-Type", contentType)
+	w.Write(rb)
+}
+
+func forwardProxyRequest(client *http.Client, targethost string, targetPath string, body []byte, contentType string) (*http.Response, error) {
+	targetURL := "https://" + targethost + targetPath
+	req, err := http.NewRequest("POST", targetURL, bytes.NewReader(body))
+	if err != nil {
+		Log.Error("Failed creating target POST request", "error", err)
+		return nil, errors.New("failed creating target POST request")
+	}
+	req.Header.Set("Content-Type", contentType)
+	return client.Do(req)
+}
+
+func (s *ODoHListener) ODoHqueryHandler(w http.ResponseWriter, r *http.Request) {
+	qHeader := r.Header.Get("Content-Type")
+	if r.Method != "POST" || qHeader == DOH_CONTENT_TYPE {
+		if s.opt.AllowDoH {
+			Log.Debug("Forwarding DoH query")
+			s.doh.dohHandler(w, r)
+			return
+		} else {
+			Log.Debug("DoH queries disabled, dropping DoH message")
+			http.Error(w, "only contentType oblivious-dns-message allowed", http.StatusMethodNotAllowed)
+			return
+		}
+	}
+
+	if qHeader != ODOH_CONTENT_TYPE {
+		http.Error(w, "only contentType oblivious-dns-message allowed", http.StatusMethodNotAllowed)
+		return
+	}
+	defer r.Body.Close()
+	b, err := io.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	msg, err := odoh.UnmarshalDNSMessage(b)
+	if err != nil {
+		http.Error(w, "error while parsing oblivious query", http.StatusBadRequest)
+		return
+	}
+
+	obliviousQuery, responseContext, err := s.odohKeyPair.DecryptQuery(msg)
+	if err != nil {
+		http.Error(w, "error while decrypting oblivious query", http.StatusBadRequest)
+		return
+	}
+
+	q := &dns.Msg{}
+	if q.Unpack(obliviousQuery.Message()) != nil {
+		http.Error(w, "unpacking oblivious query failed", http.StatusBadRequest)
+		return
+	}
+
+	a, err := s.r.Resolve(q, ClientInfo{Listener: s.id, TLSServerName: r.TLS.ServerName})
+	if err != nil {
+		Log.Error("failed to resolve", "error", err)
+		a = new(dns.Msg)
+		a.SetRcode(q, dns.RcodeServerFailure)
+	}
+
+	p, err := a.Pack()
+	if err != nil {
+		Log.Error("failed to encode response", "error", err)
+		return
+	}
+
+	response := odoh.CreateObliviousDNSResponse(p, 0)
+	obliviousResponse, err := responseContext.EncryptResponse(response)
+	if err != nil {
+		http.Error(w, "failed to create oblivious response", http.StatusBadRequest)
```

Comment:
If we fail to encrypt the response, is that a BadRequest or an InternalServerError? Maybe the client provided a bad key?